### PR TITLE
[SHORT][Code Refactoring] Move linear and fourier Operators

### PR DESCRIPTION
**Pull request by chaithyagr**
This is first PR in a series of PR for carrying out #36 . It is first part of #52

### CODE REVIEW

File: mri/operators/fourier/utils.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
+# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
+# for details.
+##########################################################################
+
+"""
+Common tools for MRI image reconstruction.
+"""
+
+
```
Comment (chaithyagr) [inline mri/operators/fourier/utils.py]: This file just holds the generate operators from previous file. I moved the previous file to mri/operators/fourier/utils as they are all utils for fourier.
@zaccharieramzi  You ideally dont have to review this file.

**Comment by chaithyagr** [comment]
@zaccharieramzi you can review this now.

File: mri/operators/fourier/non_cartesian.py [hunk 2]
```diff
@@ -0,0 +1,706 @@
+# -*- coding: utf-8 -*-
```
Comment (zaccharieramzi) [inline mri/operators/fourier/non_cartesian.py]: I think this file is the exact copy of the one above no?

File: mri/operators/utils/__init__.py [hunk 3]
```diff
@@ -7,10 +7,10 @@
...
+""" This module defines the common operators.
 """
 
-# Package import
-from mri.reconstruct.linear import WaveletN
-from mri.reconstruct.linear import WaveletUD2
+# Fourier utils
+from ..fourier.utils import convert_mask_to_locations, \
+    convert_locations_to_mask, normalize_frequency_locations, \
+    get_stacks_fourier
```
Comment (zaccharieramzi) [inline mri/operators/utils/__init__.py]: Why not also link to linear utils?

File: mri/operators/fourier/utils.py (hunk 1 above)
Comment (zaccharieramzi) [reply mri/operators/fourier/utils.py]: Yes, next time don't forget to use `git mv` if possible.

File: mri/operators/_base.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# #############################################################################
```
Comment (zaccharieramzi) [inline mri/operators/_base.py]: As discussed, we don't want to have `_base` but `base`.

File: mri/operators/linear/wavelet.py [hunk 5]
```diff
@@ -13,6 +13,7 @@
 
 
 # Package import
+from .._base import OperatorBase
```
Comment (zaccharieramzi) [inline mri/operators/linear/wavelet.py]: Watch out for the name change in `base`.

File: mri/test/test_optimizer.py [hunk 6]
```diff
@@ -48,8 +48,8 @@ def test_reconstruction_fista_fft2(self):
         print("Process test FFT FISTA::")
         for image in self.images:
             fourier = FFT(samples=convert_mask_to_locations(
-                                            fftshift(self.mask)),
```
Comment (zaccharieramzi) [inline mri/test/test_optimizer.py]: I don't know what these whitelines changes are.
Ideally we should avoid those (even if can be hidden in github review) because they clutter unnecessarily the git history.
When having to write on multiple lines, the best way to do it to avoid having to make whiteline changes is:
```python
var = long_function_call(
	var1=var1,
	var2=var2,
	var3=var3,
)
```
This will limit at best the number of unnecessary changes (whitelines and commas).

File: mri/operators/_base.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#  pySAP - Copyright (C) CEA, 2017 - 2018                                     #
+#  Distributed under the terms of the CeCILL-B license,                       #
+#  as published by the CEA-CNRS-INRIA. Refer to the LICENSE file or to        #
+#  http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html for details.   #
+# #############################################################################
+
+
+class OperatorBase(object):
+    """ Base Operator class. Every operator inherits from this class,
```
Comment (zaccharieramzi) [inline mri/operators/_base.py]: every linear operator right?

File: mri/operators/_base.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    to ensure that we have all the functions rightly implemented
+    as required by Modopt
+    """
+
+    def op(self, img):
```
Comment (zaccharieramzi) [inline mri/operators/_base.py]: not necessarily on an image.

File: mri/operators/_base.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        Returns
+        -------
+        result: np.ndarray
+            operator transform of the input.
+        """
+        raise NotImplementedError("'op' is an abstract method.")
+
+    def adj_op(self, x):
+        """ This method calculates adjoint operator transform of real or
+        complex sequence.
```
Comment (zaccharieramzi) [inline mri/operators/_base.py]: not necessarily a sequence. I don't think you need to specify that it can be applied to complex-valued or real-valued data.

File: mri/operators/utils/__init__.py (hunk 3 above)
Comment (chaithyagr) [reply mri/operators/utils/__init__.py]: Missed it, cause as of now we dont use linear utils. But will add them.

File: mri/operators/_base.py (hunk 4 above)
Comment (chaithyagr) [reply mri/operators/_base.py]: yes on it

File: mri/operators/linear/wavelet.py (hunk 5 above)
Comment (chaithyagr) [reply mri/operators/linear/wavelet.py]: yup

File: mri/test/test_optimizer.py (hunk 6 above)
Comment (chaithyagr) [reply mri/test/test_optimizer.py]: Yes, but I will not change them here as this file will change fully in future. (We will use reconstructors)

File: mri/operators/_base.py (hunk 7 above)
Comment (chaithyagr) [reply mri/operators/_base.py]: Every operator, including fourier. Missed on removing fourier base

File: mri/operators/_base.py (hunk 8 above)
Comment (chaithyagr) [reply mri/operators/_base.py]: Will make it general. It is not necessarrily anything here. Its a base template

File: mri/operators/_base.py (hunk 9 above)
Comment (chaithyagr) [reply mri/operators/_base.py]: Again it is a template.

File: mri/operators/fourier/utils.py (hunk 1 above)
Comment (chaithyagr) [reply mri/operators/fourier/utils.py]: I did and thats what caused this. What I needed was git copy . Which I dont think exists

File: mri/operators/utils/__init__.py (hunk 3 above)
Comment (zaccharieramzi) [reply mri/operators/utils/__init__.py]: I don't think you need to comment these imports, the names are pretty self-explanatory.

File: mri/operators/base.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+# #############################################################################
+
+
+class OperatorBase(object):
+    """ Base Operator class. Every operator inherits from this class,
+    to ensure that we have all the functions rightly implemented
+    as required by Modopt
+    """
+
+    def op(self, input):
```
Comment (zaccharieramzi) [inline mri/operators/base.py]: `input` is protected in Python. I think `x` or `data` are fine here.

File: mri/operators/base.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
+# #############################################################################
+#  pySAP - Copyright (C) CEA, 2017 - 2018                                     #
+#  Distributed under the terms of the CeCILL-B license,                       #
+#  as published by the CEA-CNRS-INRIA. Refer to the LICENSE file or to        #
+#  http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html for details.   #
+# #############################################################################
+
+
+class OperatorBase(object):
+    """ Base Operator class. Every operator inherits from this class,
```
Comment (zaccharieramzi) [inline mri/operators/base.py]: I don't know why you resolved (let me resolve the conversations I start for the next ones), but here I think we should specify that we are talking about linear operators (even if the notion of adjoint is larger than that).

File: mri/operators/base.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        Returns
+        -------
+        result: np.ndarray
+            operator transform of the input.
+        """
+        raise NotImplementedError("'op' is an abstract method.")
+
+    def adj_op(self, x):
+        """ This method calculates adjoint operator transform of real or
+        complex data.
```
Comment (zaccharieramzi) [inline mri/operators/base.py]: Again here I think this specification is unnecessary.

File: mri/operators/utils/__init__.py (hunk 3 above)
Comment (chaithyagr) [reply mri/operators/utils/__init__.py]: Well now that they are in, I think I will keep them there. Anyway it creates distinction in code lines for future.

File: mri/operators/base.py (hunk 10 above)
Comment (chaithyagr) [reply mri/operators/base.py]: Oh damn I forgot, yes on it.

File: mri/operators/base.py (hunk 11 above)
Comment (chaithyagr) [reply mri/operators/base.py]: Ahh, Im sorry, I missed that this had you to answer.
Hmm makes sense..

File: mri/operators/base.py (hunk 12 above)
Comment (chaithyagr) [reply mri/operators/base.py]: Well, its just an abstract class, so it doesnt really matter, but again I think ill keep it there now that it is there.

Comment (zaccharieramzi) [reply mri/operators/base.py]: Even if it's an abstract class we still want to provide accurate and efficient doc. Here this makes it seem like there is an additional catch.

File: mri/operators/fourier/utils.py [hunk 13]
```diff
@@ -272,8 +272,8 @@ def get_stacks_fourier(kspace_loc):
     sort_pos = np.lexsort(tuple(kspace_loc[:, i]
                                 for i in np.arange(3)))
     kspace_loc = kspace_loc[sort_pos]
-    first_stack_len = np.size(np.where(kspace_loc[:, 2]
-                                       == np.min(kspace_loc[:, 2])))
+    first_stack_len = np.size(np.where(
+        kspace_loc[:, 2] == np.min(kspace_loc[:, 2])))
```
Comment (zaccharieramzi) [inline mri/operators/fourier/utils.py]: Just as a nitpicking (don't need to correct here), the best way to avoid future unnecessary changes is :
```python
first_stack_len = np.size(np.where(
        kspace_loc[:, 2] == np.min(kspace_loc[:, 2]),
))
```

File: mri/operators/fourier/utils.py (hunk 1 above)
Comment (zaccharieramzi) [reply mri/operators/fourier/utils.py]: Caused what?

File: mri/operators/fourier/utils.py (hunk 13 above)
Comment (chaithyagr) [reply mri/operators/fourier/utils.py]: Yes. Duly noted, this will change in future anyway.

**Comment by chaithyagr** [comment]
@zaccharieramzi , You can merge this to mainline once everything pass. I will rebase my next set of codes and start a PR immediately/
